src/helpers.c

```diff
@@ -149,130 +149,108 @@ void concat_str_array(char **output, char *const argv[])
 	}
 }
 
-int system_argv(char *const argv[])
+int system_argv_pipe(char *const lhscmd[], char *const rhscmd[])
 {
-	int child_exit_status;
-	pid_t pid;
-	int status = -1;
-
-	pid = fork();
-
-	if (pid == 0) { /* child */
-		execvp(*argv, argv);
-		LOG(NULL, "This line must not be reached", "");
-		assert(0);
-	} else if (pid < 0) {
-		LOG(NULL, "Failed to fork a child process", "");
-		assert(0);
-	} else {
-		pid_t ws = waitpid(pid, &child_exit_status, 0);
+	return system_argv_pipe_fd(-1, -1, lhscmd, -1, -1, rhscmd);
+}
 
-		if (ws == -1) {
-			LOG(NULL, "Failed to wait for child process", "");
+int system_argv_pipe_fd(int lnewstdinfd, int lnewstderrfd, char *const lhscmd[],
+                        int rnewstdoutfd, int rnewstderrfd, char *const rhscmd[])
+{
+	pid_t monitorpid = fork();
+	if (monitorpid == -1) {
+		LOG(NULL, "Failed to create child process to monitor pipe between", "command %s and command %s", lhscmd[0], rhscmd[0]);
+		return -1;
+	} else if (monitorpid == 0) {
+		int pipefd[2];
+		if (pipe(pipefd) == -1) {
+			LOG(NULL, "Failed to create a pipe between", "command %s and command %s", lhscmd[0], rhscmd[0]);
 			assert(0);
 		}
-
-		if (WIFEXITED(child_exit_status)) {
-			status = WEXITSTATUS(child_exit_status);
-		} else {
-			LOG(NULL, "Child process didn't exit", "");
+		pid_t lhspid = system_argv_fd_nowait(-1, pipefd[1], -1 , lhscmd);
```

Comment:
I think you want
pid_t lhspid = system_argv_fd_nowait(lnewstdinfd, pipefd[1], lnewstderrfd , lhscmd); 
and a similar change at line 175. 

Note comment on line 206 means you need to pass both pipefd[0] and pipefd[1] here, and on line 175 pass pipfd[1] and pipefd[0] (i.e. reversed)

Reply:
Oops, yes, I forgot to change this.

Reply:
I might be misunderstanding the second part of the comment.  Passing both ends of the pipe on lines 170 and 175 would, in addition to the current behavior, set the STDIN for the LHS command to be the read end of the pipe, and set the STDOUT for the RHS command to be the write end of the pipe.  I didn't set these because I didn't think commands involved in a pipe should be reading or writing data in this direction.  If a chain of pipes was being made where additional redirection was needed, another pipe should be made, right?

Reply:
I think system_argv_fd_nowait requires a 5th parameter, a fd to optionally close. See comment on line 206.

(I think I am seeing how I am driving GitHub reviews incorrectly. I have multiple comments open and then press "finish review" and it doesn't warn me that these comments have not been saved).

Reply:
If I wanted to summarize our pipe file descriptor management discussions, the long and short of it is: do not keep unnecessary pipe file descriptors open in any process, forked or not, because otherwise it introduces the risk of accidental reuse.  We want a guarantee that this won't happen, so we close pipe file descriptors immediately if they are not needed before exec'ing any commands.  Being lax about pipe file descriptor management can block SIGPIPE from being delivered, or from EOF being given to a command; leading to hung processes.  Does that sound about right?

(That sounds like it was how the large comment was missed during our review on Wednesday.)

Reply:
Yes, that is a good summary.

https://www.youtube.com/watch?v=JDSPwexlyTo